src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp

```diff
@@ -332,21 +373,154 @@ void VTKPolyDataWriterInterface::writeElementFields( vtkSmartPointer< vtkCellDat
+                                                          string const & field ) const
+{
+  localIndex count = 0;
+
+  vtkSmartPointer< VTKGEOSXData > data = VTKGEOSXData::New();
+  data->SetNumberOfValues( er.getNumberOfElements< SUBREGION >() );
+  data->SetName( field.c_str() );
+
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    writeField( wrapper, data, esr.size(), count );
+  } );
+  celldata->AddArray( data );
+}
+
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementField4D( vtkSmartPointer< vtkCellData > const celldata,
+                                                      ElementRegionBase const & er,
+                                                      string const & field ) const
+{
+  // Step 1: we need to figure out the size of the two dimensions of the array that we are going to output component by component
+  localIndex nCompInSecondDim = 0;
+  localIndex nCompInThirdDim = 0;
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapperBase = esr.getWrapperBase( field );
+    std::type_info const & typeID = wrapperBase.getTypeId();
+    rtTypes::applyArrayTypeLambda2( rtTypes::typeID( typeID ),
+                                    true,
+                                    [&]( auto array, auto GEOSX_UNUSED_PARAM( Type ) )
+    {
+      typedef decltype( array ) arrayType;
+      Wrapper< arrayType > const & wrapperT = dynamicCast< Wrapper< arrayType > const & >( wrapperBase );
+      traits::ViewTypeConst< arrayType > const sourceArray = wrapperT.reference().toViewConst();
+
+      GEOSX_THROW_IF( arrayType::NDIM < 4,
+                      "An array4d is expected in this function",
+                      InputError );
+
+      nCompInSecondDim = sourceArray.size( 1 );
+      nCompInThirdDim = sourceArray.size( 2 );
+    } );
+  } );
+
+  GEOSX_THROW_IF( nCompInSecondDim == 0 || nCompInThirdDim == 0,
+                  "The array4d should have at least one component in the second and third dimensions",
+                  InputError );
+
+  // Step 2: we know how many components we have in the two dimensions that will be displayed component by component
+  //         we can now loop over the components of these dimensions and output to VTK in separate instances of VTKGEOSXData
+  for( localIndex iComp2 = 0; iComp2 < nCompInSecondDim; ++iComp2 )
+  {
+    for( localIndex iComp3 = 0; iComp3 < nCompInThirdDim; ++iComp3 )
+    {
+      localIndex count = 0;
+
+      // Step 2.1: write the name of the specific component iComp we consider here
+      vtkSmartPointer< VTKGEOSXData > data = VTKGEOSXData::New();
+      data->SetNumberOfValues( er.getNumberOfElements< SUBREGION >() );
+
+      if( nCompInSecondDim == 1 )
+      {
+        data->SetName( (field+"_"+std::to_string( iComp3 )).c_str() ); // special case for nice display of constitutive fields
```

Comment:
At some point we will need to handle constitutive fields entirely separately like Silo does... because we have to average over quadrature points - important for mechanics quantities like stress. Just a comment, this is fine for now :)